server/api/jurisdictions.py

```diff
@@ -535,6 +537,64 @@ def status(jurisdiction_id: str) -> JurisdictionStatus:
     }
 
 
+@api.route(
+    "/election/<election_id>/jurisdictions/last-login",
+    methods=["GET"],
+)
+@restrict_access([UserType.AUDIT_ADMIN])
+def get_last_login_by_jurisdiction(election: Election):
+    # Get all jurisdiction IDs for this election
+    jurisdiction_ids = [
+        jurisdiction_id
+        for (jurisdiction_id,) in Jurisdiction.query.with_entities(Jurisdiction.id)
+        .filter_by(election_id=election.id)
+        .all()
+    ]
+
+    # Look for events after most recent round, or election creation time
+    # if no rounds exist yet
+    query_timestamp_after = election.created_at
+    round = (
+        Round.query.filter_by(election_id=election.id)
+        .order_by(Round.created_at.desc())
+        .one_or_none()
+    )
+    if round is not None:
+        query_timestamp_after = round.created_at
+
+    # Query for login activities from users who are related to jurisdictions we found earlier
+    activities = (
+        ActivityLogRecord.query.join(
+            User, ActivityLogRecord.info["base"]["user_key"].astext == User.email
+        )
+        .join(User.jurisdictions)
+        .filter(
+            Jurisdiction.id.in_(jurisdiction_ids),
+            ActivityLogRecord.organization_id == election.organization_id,
+            ActivityLogRecord.timestamp > query_timestamp_after,
+            ActivityLogRecord.activity_name == "JurisdictionAdminLogin",
+        )
+        .add_columns(Jurisdiction.id.label("jurisdiction_id"))
+        .all()
+    )
+
+    # Get the most recent login activity for each jurisdiction
+    last_login_by_jurisdiction: Dict[str, ActivityLogRecord] = {}
+    for activity, jurisdiction_id in activities:
+        if (jurisdiction_id not in last_login_by_jurisdiction) or (
+            activity.timestamp > last_login_by_jurisdiction[jurisdiction_id].timestamp
+        ):
+            last_login_by_jurisdiction[jurisdiction_id] = activity
```

Comment:
It's possible we could bake this postprocessing into the query, but I'm not sure it would be more readable

Reply:
I think you can do it like so:
```py
 .order_by(Jurisdiction.id, ActivityLogRecord.timestamp.desc())
 .distinct(Jurisdiction.id)
 .with_entities(Jurisdiction.id, ActivityLogRecord)
```

This is nice because then you can just wrap the whole query expression with `dict()` and that will be `last_login_by_jurisdiction`

Reply:
https://github.com/votingworks/arlo/pull/2061/commits/1552349d670efb58c45844de06fda8cea005718d